tests/unit/lib/charms/sdcore_nms_k8s/v0/test_fiveg_core_gnb_provider_interface.py

```diff
@@ -0,0 +1,171 @@
+                            "type": "string",
+                        },
+                        "tac": {
+                            "type": "string",
+                        },
+                        "plmns": {
+                            "type": "string",
+                        },
+                    },
+                },
+                "publish-gnb-config-wrong-data": {
+                    "params": {
+                        "relation-id": {
+                            "description": "The relation ID of the relation",
+                            "type": "string",
+                        },
+                        "tac": {
+                            "type": "string",
+                        },
+                        "plmns": {
+                            "type": "string",
+                        },
+                    },
+                },
+            },
+        )
+
+    @pytest.fixture(autouse=True)
+    def setUp(self, request):
+        yield
+        request.addfinalizer(self.tearDown)
+
+    def tearDown(self) -> None:
+        patch.stopall()
+
+    def test_given_fiveg_core_gnb_relation_when_publish_gnb_config_valid_tac_then_data_is_in_application_databag(  # noqa: E501
+        self,
+    ):
+        fiveg_core_gnb_relation = scenario.Relation(
+            endpoint="fiveg_core_gnb",
+        )
+        state_in = scenario.State(
+            leader=True,
+            relations={fiveg_core_gnb_relation},
+        )
+
+        plmns = [PLMNConfig(mcc=TEST_MCC, mnc=TEST_MNC, sst=TEST_SST, sd=TEST_SD)]
+        params = {
+            "relation-id": str(fiveg_core_gnb_relation.id),
+            "tac": str(TEST_TAC_VALID),
+            "plmns": json.dumps([plmn.asdict() for plmn in plmns])
+        }
+
+        state_out = self.ctx.run(self.ctx.on.action("publish-gnb-config", params=params),
+                                 state_in)
+        assert (
+            state_out.get_relation(fiveg_core_gnb_relation.id).local_app_data["tac"]
+            == str(TEST_TAC_VALID)
+        )
+        rel_plmns = state_out.get_relation(fiveg_core_gnb_relation.id).local_app_data["plmns"]
+        assert plmns == [PLMNConfig(**data) for data in json.loads(rel_plmns)]
+
+    def test_given_fiveg_core_gnb_relation_when_publish_gnb_config_invalid_tac_then_exception_is_raised(  # noqa: E501
+        self,
+    ):
+        fiveg_core_gnb_relation = scenario.Relation(
+            endpoint="fiveg_core_gnb",
+        )
+        state_in = scenario.State(
+            leader=True,
+            relations={fiveg_core_gnb_relation},
+        )
+
+        plmns = [PLMNConfig(mcc=TEST_MCC, mnc=TEST_MNC, sst=TEST_SST, sd=TEST_SD)]
+        params = {
+            "relation-id": str(fiveg_core_gnb_relation.id),
+            "tac": str(TEST_TAC_INVALID),
+            "plmns": json.dumps([plmn.asdict() for plmn in plmns])
+        }
+
+        with pytest.raises(Exception):
```

Comment:
I know that setting a specific exception here fails, but I think that you can check the text in the exception (`Invalid fiveG core gNB data`) to make sure that it failed in the library when calling `publish_gnb_config_information` and not during the transformation from string to actual type during test setup.
Same comment for all tests that raise an exception.